src/librustc_infer/traits/wf.rs

```diff
@@ -323,8 +323,15 @@ impl<'a, 'tcx> WfPredicates<'a, 'tcx> {
             let implied_obligations = traits::elaborate_obligations(tcx, obligations.clone());
             let implied_obligations = implied_obligations.map(|obligation| {
                 let mut cause = cause.clone();
-                cause.code =
-                    traits::ObligationCauseCode::DerivedObligation(Rc::new(obligation.cause.code));
+                let parent_trait_ref = obligation
+                    .predicate
+                    .to_opt_poly_trait_ref()
+                    .unwrap_or_else(|| ty::Binder::dummy(*trait_ref));
+                let derived_cause = traits::DerivedObligationCause {
+                    parent_trait_ref,
+                    parent_code: Rc::new(obligation.cause.code.clone()),
+                };
+                cause.code = traits::ObligationCauseCode::ImplDerivedObligation(derived_cause);
                 extend_cause_with_original_assoc_item_obligation(
```

Comment:
Maybe `extend_cause_with_original_assoc_item_obligation` should use a derived cause *only* if it doesn't apply an associated type bound cause?